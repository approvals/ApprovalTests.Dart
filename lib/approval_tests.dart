import 'dart:convert';
import 'dart:io';
import 'dart:math';
import 'dart:mirrors';

import 'package:dcli/dcli.dart';
import 'package:talker/talker.dart';

part 'src/approvals.dart';
part 'src/writers/approval_text_writer.dart';
part 'src/core/logger/logger.dart';
part 'src/core/utils/utils.dart';
part 'src/core/options.dart';
part 'src/namer/namer.dart';
part 'src/comparator/comparator.dart';
part 'src/core/enums/comporator_ide.dart';
part 'src/comparator/command_line_comparator.dart';
part 'src/comparator/ide_comparator.dart';
part 'src/reporter/doesnt_match_exception.dart';
part 'src/core/utils/extensions.dart';
part 'src/core/utils/converter.dart';
part 'src/core/utils/executable_query.dart';
part 'src/namer/approval_number.dart';
