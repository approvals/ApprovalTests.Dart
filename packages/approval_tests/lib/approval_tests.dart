// ignore_for_file: implementation_imports

import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:diff_match_patch2/diff_match_patch.dart';
import 'package:talker/talker.dart';
import 'package:test_api/src/backend/invoker.dart' show Invoker;

part 'src/approvals.dart';
part 'src/comparator/file_comparator.dart';
part 'src/core/approval_number.dart';
part 'src/core/approval_writer.dart';
part 'src/core/comparator.dart';
part 'src/core/enums/comporator_ide.dart';
part 'src/core/extensions/approval_string_extensions.dart';
part 'src/core/utils/file_path_extractor.dart';
part 'src/core/logger/logger.dart';
part 'src/core/options.dart';
part 'src/core/platform_wrapper.dart';
part 'src/core/reporter.dart';
part 'src/core/scrubber.dart';
part 'src/core/stack_trace_fetcher/fetcher.dart';
part 'src/core/stack_trace_fetcher/stack_trace_fetcher.dart';
part 'src/core/utils/converter.dart';
part 'src/core/utils/executable_query.dart';
part 'src/core/utils/utils.dart';
part 'src/exceptions/doesnt_match_exception.dart';
part 'src/exceptions/file_not_found_exception.dart';
part 'src/exceptions/no_diff_tool_exception.dart';
part 'src/namer/file_namer_options.dart';
part 'src/namer/namer.dart';
part 'src/reporters/command_line/command_line_reporter.dart';
part 'src/reporters/diff_tool/diff_info.dart';
part 'src/reporters/diff_tool/diff_tool_reporter.dart';
part 'src/reporters/diff_tool/diff_tools.dart';
part 'src/scrubbers/date_scrubber.dart';
part 'src/scrubbers/nothing_scrubber.dart';
part 'src/scrubbers/reg_exp_scrubber.dart';
part 'src/writers/approval_text_writer.dart';
part 'src/reporters/git.dart';
