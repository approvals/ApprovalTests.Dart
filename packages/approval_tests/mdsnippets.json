{
  "$schema": "https://raw.githubusercontent.com/SimonCropp/MarkdownSnippets/master/schema.json",
  "Convention": "InPlaceOverwrite",
  "TocLevel": 5
}
