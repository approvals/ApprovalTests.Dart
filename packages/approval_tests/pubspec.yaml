#  * 1.0.0 for a first major release.
#  * 0.1.0 for a breaking release.
#  * 0.0.4 for a minor release.
name: approval_tests
description: Approval Tests implementation in Dart. Inspired by ApprovalTests.
version: 1.0.0
repository: https://github.com/approvals/ApprovalTests.Dart
homepage: https://github.com/approvals/ApprovalTests.Dart.git
issue_tracker: https://github.com/approvals/ApprovalTests.Dart/issues

topics:
  - test
  - approval
  - approvals
  - tests
  - approval-tests

environment:
    sdk: '>=3.0.0 <4.0.0'

dependencies:
  test: ^1.25.2
  talker: ^4.3.2
  diff_match_patch2: ^0.5.0
  test_api: ^0.7.0

dev_dependencies:
  sizzle_lints: ^2.0.2