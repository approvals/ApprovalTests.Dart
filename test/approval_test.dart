// Import third-party packages
import 'package:approval_tests/approval_tests.dart';
import 'package:test/test.dart';

// Import project-specific dependencies
import 'models/item.dart';
import 'queries/db_request_query.dart';

part 'utils/helper.dart';
part 'constants/lines.dart';

void main() {
  /// ================== Init fields ==================

  const helper = ApprovalTestHelper();
  const dbQuery = DatabaseRequestQuery("1");
  const lines25 = _Lines.lines25;
  const lines30 = _Lines.lines30;

  /// ================== Set up ==================

  setUpAll(() {
    ApprovalLogger.warning("$lines30 Tests are starting $lines30");
  });

  /// ================== Approvals: verify methods ==================

  group('Approvals: verify methods |', () {
    setUpAll(() {
      ApprovalLogger.log(
        "$lines25 Group: Test verify methods are starting $lines25",
      );
    });
    test('verify string with approved result options', () {
      helper.verify(
        'Hello World',
        'verify',
      );
    });

    test('Verify all strings in a list', () {
      helper.verifyAll(['Hello World', 'Hello World'], 'verify_all');
    });

    test('Verify JSON data', () {
      helper.verifyAsJson({"message": "Hello World"}, 'verify_as_json');
    });

    test('Verify JSON data: with model', () {
      helper.verifyAsJson(ApprovalTestHelper.jsonItem, 'verify_as_model_json');
    });

    test('Verify all combinations', () {
      helper.verifyAllCombinations(
        [
          [1, 2],
          [3, 4],
        ],
        'verify_all_combinations',
      );
    });

    test("Verify sequence", () {
      helper.verifySequence([1, 2, 3], 'verify_sequence');
    });

    test("Verify query result", () async {
      await helper.verifyQuery(dbQuery, 'verify_query');
    });
  });

  /// ================== Approvals: test for exceptions ==================
  group('Approvals: test for exceptions |', () {
    setUpAll(() {
      ApprovalLogger.log(
        "$lines25 Group: Exception cases are starting $lines25",
      );
    });
    // test('Method «verify» must throw PathNotFoundException', () {
    //   expect(
    //     () => helper.verify(
    //       'Hello World',
    //       'verify_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );
    //   ApprovalLogger.success(
    //     "Test Passed: Method «verify» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    // test('Method «verifyAll» must throw PathNotFoundException', () {
    //   expect(
    //     () => helper.verifyAll(
    //       ['Hello World', 'Hello World'],
    //       'verify_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );

    //   ApprovalLogger.success(
    //     "Test Passed: Method «verifyAll» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    // test('Method «verifyAsJson» must throw PathNotFoundException', () {
    //   expect(
    //     () => helper.verifyAsJson(
    //       {"message": "Hello World"},
    //       'verify_as_json_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );

    //   ApprovalLogger.success(
    //     "Test Passed: Method «verifyAsJson» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    // test('Method «verifyAllCombinations» must throw PathNotFoundException', () {
    //   expect(
    //     () => helper.verifyAllCombinations(
    //       [
    //         [1, 2],
    //         [3, 4],
    //       ],
    //       'verify_all_combinations_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );

    //   ApprovalLogger.success(
    //     "Test Passed: Method «verifyAllCombinations» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    // test("Method «verifySequence» must throw PathNotFoundException", () {
    //   expect(
    //     () => helper.verifySequence(
    //       [1, 2, 3],
    //       'verify_sequence_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );

    //   ApprovalLogger.success(
    //     "Test Passed: Method «verifySequence» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    // test("Method «verifyQuery» must throw PathNotFoundException", () async {
    //   expect(
    //     () => helper.verifyQuery(
    //       dbQuery,
    //       'verify_query_exception',
    //       expectException: true,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );

    //   ApprovalLogger.success(
    //     "Test Passed: Method «verifyQuery» correctly throws PathNotFoundException as expected.",
    //   );
    // });

    test("Method «verify» must throw DoesntMatchException", () {
      expect(
        () => helper.verify(
          'Hello W0rld',
          'verify',
          expectException: true,
        ),
        throwsA(isA<DoesntMatchException>()),
      );

      ApprovalLogger.success(
        "Test Passed: Method «verify» correctly throws DoesntMatchException as expected.",
      );
    });

    test("Method «verify» must throw DoesntMatchException with error handling", () {
      expect(
        () => helper.verify(
          'Hello W0rld',
          'verify',
        ),
        throwsA(isA<DoesntMatchException>()),
      );
      ApprovalLogger.success(
        "Test Passed: Successfully handled a log mismatch. Method «verify» correctly throws DoesntMatchException as expected.",
      );
    });
  });

  /// ================== Approvals: test for exceptions ==================

  group('Approvals: test of other minor things |', () {
    setUpAll(() {
      ApprovalLogger.log("$lines25 Group: Minor cases are starting $lines25");
    });
    // test('Verify method without initial path: PathNotFoundException', () {
    //   expect(
    //     () => helper.verify(
    //       'Hello World',
    //       'verify_exception',
    //       expectException: true,
    //       useDefaultPath: false,
    //     ),
    //     throwsA(isA<PathNotFoundException>()),
    //   );
    //   ApprovalLogger.success(
    //     "Test Passed: The method was successfully verified for absence of an initial path — PathNotFoundException was thrown.",
    //   );
    // });

    test('Simulate file not found error during comparison. Must throw CommandLineComparatorException.', () async {
      const comparator = CommandLineComparator();

      // Setup: paths to non-existent files
      const nonExistentApprovedPath = 'path/to/nonexistent/approved.txt';
      const nonExistentReceivedPath = 'path/to/nonexistent/received.txt';

      // Expect an exception to be thrown
      expect(
        () => comparator.compare(
          approvedPath: nonExistentApprovedPath,
          receivedPath: nonExistentReceivedPath,
        ),
        throwsA(isA<CommandLineComparatorException>()),
      );

      ApprovalLogger.success(
        "Test Passed: Successfully handled a file not found error during comparison.",
      );
    });

    test('Simulate file not found error during comparison. Must throw IDEComparatorException.', () async {
      const comparator = IDEComparator();

      // Setup: paths to non-existent files
      const nonExistentApprovedPath = 'path/to/nonexistent/approved.txt';
      const nonExistentReceivedPath = 'path/to/nonexistent/received.txt';

      // Expect an exception to be thrown
      expect(
        () => comparator.compare(
          approvedPath: nonExistentApprovedPath,
          receivedPath: nonExistentReceivedPath,
        ),
        throwsA(isA<IDEComparatorException>()),
      );

      ApprovalLogger.success(
        "Test Passed: Successfully handled a file not found error during comparison.",
      );
    });

    test('verify string with IDEComparator', () {
      expect(
        () => helper.verify(
          'Hello W0rld',
          'verify',
          deleteReceivedFile: false,
          comparator: const IDEComparator(),
        ),
        throwsA(isA<DoesntMatchException>()),
      );

      ApprovalLogger.success(
        "Test Passed: Successfully handled a log mismatch. Method «verify» correctly throws DoesntMatchException as expected.",
      );
    });

    test('Verify string with scrubber', () {
      helper.verify(
        '  Hello    World  \t\n ',
        'verify_scrub',
        scrubber: const ScrubWithRegEx(),
      );
    });
  });

  /// ================== Tear down ==================

  tearDownAll(() {
    ApprovalLogger.warning("$lines30 All tests are done $lines30");
  });
}
