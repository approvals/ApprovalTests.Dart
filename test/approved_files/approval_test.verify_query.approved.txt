# This file was generated by approval_tests. Please do not edit.

{"id": "1", "name": "John Doe", "email": "john@example.com"}